### DISCUSSION TITLE
runtime-portable

### DISCUSSION THREAD

**Discussion by tmds** [dotnet/source-build#3012] [Q&A] [answered]
I'm building our Fedora .NET 6 packages. I don't do it often (enough).

One thing that stands-out for me is that it is doing two runtime builds, one called `runtime-portable` and one called `runtime`.

I assume the packages in `dotnet-sdk-6.0-source-built-artifacts` were already built from source.
So, why are we doing two passes?

cc @crummel @MichaelSimons @omajid

**Comment by tmds** [comment]
Who can answer my questions?

Does our `dotnet-sdk-6.0-source-built-artifacts` package contain pre-builts from Microsoft or is it all built-from-source?

Why are we doing both a portable and a non-portable rid? Is it to get rid of pre-builts? Or are there other reasons?

How does source-build manage the dependencies/order between the different repos? Is it the order in the `known-good.proj`?

cc @crummel

**Comment by omajid** [comment]
> Does our dotnet-sdk-6.0-source-built-artifacts package contain pre-builts from Microsoft or is it all built-from-source?

It should be everything built from source (only). We use [poisoning](https://github.com/dotnet/source-build/pull/797) to check that unwanted files don't leak through.

> Why are we doing both a portable and a non-portable rid? Is it to get rid of pre-builts? Or are there other reasons?

IIRC, it is needed for bootstrapping and a few scenarios: https://github.com/dotnet/source-build/pull/1433. In 5.0 (I think) we used to build non-portable only and that seemed to work. Perhaps we should re-evaluate this? Or provide post-bootstrap guidance for distros?

> How does source-build manage the dependencies/order between the different repos? Is it the order in the known-good.proj?

AFAIK, the build order is in known-good.proj, but the actual product dependencies are tracked in eng/Version.Details.xml in each repository. I have found https://devblogs.microsoft.com/dotnet/a-deep-dive-into-how-net-builds-and-ships/ to be a good overview of the file and its use.

**Comment by tmds** [comment]
> In 5.0 (I think) we used to build non-portable only and that seemed to work.

Yes, this is what I expect.

We shouldn't have to build the runtime twice if we're starting from all source-built artifacts (`dotnet-sdk-6.0-source-built-artifacts` package).

> he build order is in known-good.proj

Several things get built using the pre-built bootstrap sdk ([built with bootstrap sdk](https://github.com/dotnet/installer/blob/ed245c939279170f709c46b1920b1f3e6054f07d/src/SourceBuild/tarball/content/repos/known-good.proj#L28-L63)) and bootstrap runtime ([built with bootstrap runtime](https://github.com/dotnet/installer/blob/ed245c939279170f709c46b1920b1f3e6054f07d/src/SourceBuild/tarball/content/repos/known-good.proj#L28-L46)).

Why are we building the runtime twice, but none of the other stuff?

**Comment by crummel** [comment]
At one time, ASP.NET had a hard requirement on the portable runtime assets.  This was something they intended to fix, I'll dig up the bug.  If it's been fixed then yes, I think that's the only blocker to getting rid of always building runtime-portable.  Let me do some checking on this.

**Comment by tmds** [comment]
I've looked closer at the aspnetcore build, and made a PR that allows to use the non-portable runtime packages: https://github.com/dotnet/aspnetcore/pull/43937.

With that change, we no longer need to build `runtime-portable`.

source-build needs to set the `/p:TargetRuntimeIdentifier=$(TargetRid)` when building aspnetcore.
We also need to update the name of the ASP.NET Core tarballs used by the `installer` repo.

**Comment by tmds** [comment, marked as answer]
There is more to it than ASP.NET Core. I've created https://github.com/dotnet/source-build/issues/3027 to track things needed to remove the `runtime-portable` build.
